android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt

```diff
@@ -8,109 +8,237 @@ import android.graphics.Rect
-    val fitsDown = availableHeightToBottom >= containerHeight
-    val reversed = availableWidthToLeft < availableWidthToRight
+    if (recyclerViewAdapter == null) {
+        Log.log(
+            priority = Log.Priority.ERROR,
+            tag = "mozac-popupwindow",
+            message = "The browser menu recyclerview has no adapter set.",
+        )
+        return false
+    }
+
+    val containerViewHeight = containerView.measuredHeight
+    val recyclerViewHeight = containerRecyclerView.measuredHeight
+    val recyclerViewItemCount = recyclerViewAdapter.itemCount
+
+    val hasViewInitializedCorrectly =
+        containerViewHeight > 0 && recyclerViewHeight > 0 && recyclerViewItemCount > 0
+    if (!hasViewInitializedCorrectly) {
+        Log.log(
+            priority = Log.Priority.ERROR,
+            tag = "mozac-popupwindow",
+            message = "containerViewHeight: $containerViewHeight must be more than 0," +
+                "recyclerViewHeight: $recyclerViewHeight must be more than 0 &" +
+                "recyclerViewItemCount: $recyclerViewItemCount must be more than 0.",
+        )
+        return false
+    }
 
     // Popup window does not need a input method. This avoids keyboard flicker when menu is opened.
     inputMethodMode = PopupWindow.INPUT_METHOD_NOT_NEEDED
 
-    // Try to use the preferred orientation, if doesn't fit fallback to the best fit.
-    when {
-        preferredOrientation == Orientation.DOWN && (fitsDown || forceOrientation) ->
-            showPopupWithDownOrientation(anchor, reversed)
-        preferredOrientation == Orientation.UP && (fitsUp || forceOrientation) ->
-            showPopupWithUpOrientation(anchor, containerHeight, reversed)
+    showPopup(orientation, anchor, recyclerViewHeight, recyclerViewItemCount, containerViewHeight)
+    return true
+}
+
+/**
+ * Show the [PopupWindow] with the required configuration.
+ */
+private fun PopupWindow.showPopup(
+    orientation: Orientation?,
+    anchor: View,
+    recyclerViewHeight: Int,
+    recyclerViewItemCount: Int,
+    containerViewHeight: Int,
+) {
+    val containerHeight =
+        getContainerHeight(recyclerViewHeight, recyclerViewItemCount, containerViewHeight, anchor)
+
+    val (availableWidthToLeft, availableWidthToRight) = getMaxAvailableWidthToLeftAndRight(anchor)
+    val isRTL = availableWidthToLeft > availableWidthToRight
+
+    when (orientation) {
+        Orientation.DOWN -> showPopupWithDownOrientation(anchor, containerHeight, isRTL)
+
+        Orientation.UP -> {
+            val availableHeightToTop = getMaxAvailableHeightToTopAndBottom(anchor).first
+
+            // This could include cases where the keyboard is being displayed.
+            if (containerHeight > availableHeightToTop) {
+                showPopupWhereBestFits(anchor, isRTL, containerHeight)
+            } else {
+                showPopupWithUpOrientation(anchor, containerHeight, isRTL)
+            }
+        }
+
         else -> {
```

Comment:
> Maybe using view.isRTL inside PopupWindow.displayPopup would be a better approach? It would actually tell the direction of the layout without relying on the start/end position of the anchor

Good suggestion, I don't see any issue with this (@rocketsroger thoughts?) I'll update to use the view.